Tests/src/InternalTests/CommonTypeTests/Optional.cpp

```diff
@@ -0,0 +1,290 @@
+		EXPECT_EQ(&(*CopyOptionalInstance), nullptr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithExtendedClassReferenceInitialisationTest)
+{
+	try
+	{
+		OptionalExtraTestClass TestClassInstance(5, 6);
+		Optional<OptionalTestClass> OptionalInstance(TestClassInstance);
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstance.SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstance.SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithValueInitialisationTest)
+{
+	try
+	{
+		OptionalTestClass TestClassInstance(7);
+		Optional<OptionalTestClass> OptionalInstance(TestClassInstance);
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstance.SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstance.SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithMoveValueInitialisationTest)
+{
+	try
+	{
+		OptionalTestClass TestClassInstance(8);
+		Optional<OptionalTestClass> OptionalInstance(std::move(TestClassInstance));
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstance.SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstance.SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, CopyOptionalWithValueInitialisationTest)
+{
+	try
+	{
+		OptionalTestClass TestClassInstance(9);
+		Optional<OptionalTestClass> OriginalOptionalInstance(TestClassInstance);
+
+		EXPECT_TRUE(OriginalOptionalInstance.HasValue());
+		EXPECT_EQ(OriginalOptionalInstance->CopyCount, 1);
+
+		Optional<OptionalTestClass> CopyOptionalInstance(OriginalOptionalInstance);
+
+		EXPECT_TRUE(CopyOptionalInstance.HasValue());
+
+		EXPECT_EQ(OriginalOptionalInstance->CopyCount, 1);
+		EXPECT_EQ(CopyOptionalInstance->CopyCount, 2);
+
+		EXPECT_EQ((*CopyOptionalInstance).SomeField, TestClassInstance.SomeField);
+		EXPECT_EQ(CopyOptionalInstance->SomeField, TestClassInstance.SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, MoveOptionalWithValueInitialisationTest)
+{
+	try
+	{
+		OptionalTestClass TestClassInstance(10);
+		Optional<OptionalTestClass> OriginalOptionalInstance(TestClassInstance);
+		Optional<OptionalTestClass> MoveOptionalInstance(std::move(OriginalOptionalInstance));
+
+		EXPECT_TRUE(MoveOptionalInstance.HasValue());
+		EXPECT_EQ((*MoveOptionalInstance).SomeField, TestClassInstance.SomeField);
+		EXPECT_EQ(MoveOptionalInstance->SomeField, TestClassInstance.SomeField);
+
+		EXPECT_TRUE(OriginalOptionalInstance->Moved);
+		EXPECT_FALSE(MoveOptionalInstance->Moved);
+		EXPECT_EQ(MoveOptionalInstance->MoveCount, 1);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+#endif
```

Comment:
Can you add some tests for the operators we overload?

Reply:
Will do.